src/plugins/midi-piano/midi-piano-display.js

```diff
@@ -0,0 +1,292 @@
+    setTimeout(() => {
+      sampler.current.triggerRelease(getNoteNameFromMidiValue(midiValue));
+    }, 200);
+  };
+
+  const getData = () => {
+    try {
+      axios.get(src, { responseType: 'arraybuffer' })
+        .then(response => {
+          midiData.current = response.data;
+        });
+    } catch (error) {
+      // eslint-disable-next-line no-console
+      console.log(error);
+    }
+  };
+
+  const renderControls = () => (
+    <div style={{ display: 'flex', justifyContent: 'center', paddingBottom: '1rem' }}>
+      <Button onClick={startMidiPlayer} icon={<PlayIcon />} />
+      <Button onClick={pauseMidiPlayer} icon={<PauseIcon />} />
+      <Button onClick={stopMidiPlayer} icon={<QuestionCircleOutlined />} />
+    </div>
+  );
+
+  const renderMidiTrackTitle = () => (
+    <div style={{ display: 'flex', justifyContent: 'center' }}>{midiTrackTitle}</div>
+  );
+
+  useEffect(() => {
+    navigator.requestMIDIAccess().then(onMIDISuccess, onMIDIFailure);
+  });
+
+  useEffect(() => {
+    if (sampler.current) {
+      return;
+    }
+    // eslint-disable-next-line no-console
+    console.log('instantiating sampler');
+    sampler.current = new Tone.Sampler({
+      urls: {
+        'A0': 'A0.mp3',
+        'C1': 'C1.mp3',
+        'D#1': 'Ds1.mp3',
+        'F#1': 'Fs1.mp3',
+        'A1': 'A1.mp3',
+        'C2': 'C2.mp3',
+        'D#2': 'Ds2.mp3',
+        'F#2': 'Fs2.mp3',
+        'A2': 'A2.mp3',
+        'C3': 'C3.mp3',
+        'D#3': 'Ds3.mp3',
+        'F#3': 'Fs3.mp3',
+        'A3': 'A3.mp3',
+        'C4': 'C4.mp3',
+        'D#4': 'Ds4.mp3',
+        'F#4': 'Fs4.mp3',
+        'A4': 'A4.mp3',
+        'C5': 'C5.mp3',
+        'D#5': 'Ds5.mp3',
+        'F#5': 'Fs5.mp3',
+        'A5': 'A5.mp3',
+        'C6': 'C6.mp3',
+        'D#6': 'Ds6.mp3',
+        'F#6': 'Fs6.mp3',
+        'A6': 'A6.mp3',
+        'C7': 'C7.mp3',
+        'D#7': 'Ds7.mp3',
+        'F#7': 'Fs7.mp3',
+        'A7': 'A7.mp3',
+        'C8': 'C8.mp3'
+      },
+      onload: () => {
+        samplerHasLoaded.current = true;
+      },
+      baseUrl: 'https://tonejs.github.io/audio/salamander/' // Samples better be hosted in project.
+    }).toDestination();
+  });
+
+  useEffect(() => {
+    if (!src) {
+      return;
+    }
+    getData();
+  }, []);
+
+  useEffect(() => {
+    return function cleanup() {
+      if (!midiAccessObj.current) {
+        return;
+      }
+      if (player.current !== null && player.current.isPlaying()) {
+        stopMidiPlayer();
+      }
+      for (const input of midiAccessObj.current.inputs.values()) {
+        input.onmidimessage = null;
+      }
+    };
+  });
+
+  return (
+    <React.Fragment>
+      <PianoComponent
+        noteRange={noteRange}
+        playNote={playNote}
+        stopNote={stopNote}
+        keys={keys}
+        />
+      <div style={{ paddingTop: '1.5rem' }}>
+        {hasMidiFile
+          && renderControls()}
+        {hasMidiTrackTitle && hasMidiFile
```

Comment:
Small thing, since these 2 variables are used just here and the expressions they evaluate are very basic, you can reduce the code by inlining the variable expressions:
```
{!!sourceUrl && renderControls()}
{!!sourceUrl && !!midiTrackTitle && renderMidiTrackTitle() }
```
Not that less code is always better, but here it's understandable without the variables :)

Reply:
Great!